eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/CompositeRichIterable.java

```diff
@@ -0,0 +1,182 @@
+import org.eclipse.collections.impl.utility.Iterate;
+
+@Immutable
+public final class CompositeRichIterable<E>
+        extends AbstractLazyIterable<E>
+{
+    private final MutableList<RichIterable<E>> iterables;
+
+    private CompositeRichIterable(MutableList<RichIterable<E>> newIterables)
+    {
+        this.iterables = newIterables;
+    }
+
+    public CompositeRichIterable()
+    {
+        this(Lists.mutable.empty());
+    }
+
+    public static <T> CompositeRichIterable<T> with(RichIterable<T>... iterables)
+    {
+        return new CompositeRichIterable<>(Lists.mutable.with(iterables));
+    }
+
+    @Override
+    public void each(Procedure<? super E> procedure)
+    {
+        this.iterables.each(iterable -> iterable.each(procedure));
+    }
+
+    @Override
+    public void forEachWithIndex(ObjectIntProcedure<? super E> objectIntProcedure)
+    {
+        Counter index = new Counter();
+        this.iterables.each(iterable -> Iterate.forEach(iterable, object -> {
+            objectIntProcedure.value(object, index.getCount());
+            index.increment();
+        }));
+    }
+
+    @Override
+    public <P> void forEachWith(Procedure2<? super E, ? super P> procedure, P parameter)
+    {
+        this.iterables.each(iterable -> Iterate.forEachWith(iterable, procedure, parameter));
+    }
+
+    @Override
+    public boolean anySatisfy(Predicate<? super E> predicate)
+    {
+        return this.iterables.anySatisfy(each -> Iterate.anySatisfy(each, predicate));
+    }
+
+    @Override
+    public boolean allSatisfy(Predicate<? super E> predicate)
+    {
+        return this.iterables.allSatisfy(each -> Iterate.allSatisfy(each, predicate));
+    }
+
+    @Override
+    public boolean noneSatisfy(Predicate<? super E> predicate)
+    {
+        return this.iterables.noneSatisfy(each -> Iterate.anySatisfy(each, predicate));
+    }
+
+    @Override
+    public E detect(Predicate<? super E> predicate)
+    {
+        for (int i = 0; i < this.iterables.size(); i++)
+        {
+            Iterable<E> eachIterable = this.iterables.get(i);
+            E result = Iterate.detect(eachIterable, predicate);
+            if (result != null)
+            {
+                return result;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Optional<E> detectOptional(Predicate<? super E> predicate)
+    {
+        for (int i = 0; i < this.iterables.size(); i++)
+        {
+            Iterable<E> eachIterable = this.iterables.get(i);
+            Optional<E> result = Iterate.detectOptional(eachIterable, predicate);
+            if (result.isPresent())
+            {
+                return result;
+            }
+        }
+        return Optional.empty();
+    }
+
+    public void add(RichIterable<E> iterable)
+    {
+        this.iterables.add(iterable);
+    }
+
+    @Override
+    public int size()
+    {
+       return (int) this.iterables.sumOfInt(RichIterable::size);
```

Comment:
When you move this method to CompositeIterable, you'll have to replace RichIterable::size with iterable -> Iterate.sizeOf(iterable). The cast shows that we actually have a bug - you can concatenate two Intervals of size MAX_INTEGER and easily overflow the size without getting any exception.